### DISCUSSION TITLE
Search Grammar

### DISCUSSION THREAD

**Discussion by BLR-0118** [ONDC-Official/protocol-network-extension#51] [General]
ONDC base protocol has the capability for free text search. To enable free text search, a generic cross-domain search grammar (BNF/CNF/?) needs to be defined and enabled. Requesting inputs from participants for help in defining this grammar.

**Comment by BLR-0118** [comment]
@Tushar-Naik @amitsaluja @abaweja @batramanuj @EkanshKr @mahoriR @mherle @NajeebMohammed @nitinmish @nikitagupta5566 @get2abhi https://github.com/orgs/ONDC-Official/teams/participants

**Comment by nitinmish** [comment]
Notational form is only a small cog in the whole search scheme. It is generally used by mostly front end technologies to attribute intent to a keyword that has been/is being typed. The command issued through must be parsed with search libraries, irrespective whether using BNF/CNF/CFGs etc and search query is issued to a search service. Search service internally convert the query to multiple lookups to cache and extract matching entities and return back for rendering. However, the search indexes are pretty complex set of algorithms and not all platforms will have the digital maturity to handle it at that level. 

My suggestion, instead of implementing search grammar on network APIs, we should simplify it through a simple tag that carries the typed keyword as root in simpler format (may be simpler implementation of BNF being more prevalent) besides some derived keywords may also mean something else. To illustrate for better understanding, lets say a buyer typed keyword1+keyword2 on free text field, then a BNF request shall look like:

Name ::= “<domain and category identified that gets the rule context>”
keyword1 :== keyword3 (Could also be REGEX enabled) | keyword4 |keyword5|….
Keyword2 ::= string1|string2
Keyword3 ::= string3|string4
…..
…..

In this case, 
* Name will be identified by a rule getting kicked, for example in case of network, the search context will get limited to the domain and category to limit the search to identified population. It can be extended to all 4 levels identified such as RET101010AA etc.
* <keyword> could be the keyword as typed by the person performing search. This is where Buyer App must have identified possible keywords buyer may have typed and create a possible search in BNF format as mentioned above
* The seller receiving the search intent as part of /search call will extract and manifest the search BNF into further lookup able keywords and query their own catalog index, or simply pass on the BNF element to a search and discovery service provider, such as that of Google.

I hope my understanding of the problem statement is correct and coherent.

@BLR-0118
